packages/pluggableWidgets/accordion-web/src/utils/AccordionGroupStateReducer.ts

```diff
@@ -0,0 +1,41 @@
+import { AccGroup } from "../components/AccordionGroup";
+
+export type AccordionGroupsReducerAction =
+    | { type: "sync"; groups: AccGroup[] }
+    | { type: "expand" | "collapse"; group: AccGroup };
+
+export function getAccordionGroupsReducer(
+    expand: "single" | "multiple" | "all"
+): (state: AccGroup[], action: AccordionGroupsReducerAction) => AccGroup[] {
+    return (state: AccGroup[], action: AccordionGroupsReducerAction): AccGroup[] => {
+        if (action.type === "sync" || expand === "single") {
+            const newState = action.type === "sync" ? action.groups : state;
+
+            return newState.map((group, index) => {
+                let collapsed;
+
+                if (action.type === "sync") {
+                    collapsed = state[index].collapsed; // The previous collapsed state of the group at the same index can be used, because the order of groups remains the same throughout the lifetime of the widget.
+                } else {
+                    collapsed = action.type === "expand" ? group !== action.group : true;
+                }
+
+                return {
+                    ...group,
+                    collapsed
+                };
+            });
+        }
+
+        const newState = [...state];
```

Comment:
This can just be `const newState = state` or not?

Reply:
The state needs to be a new array in order to be picked up ([React docs](https://reactjs.org/docs/hooks-reference.html#bailing-out-of-a-dispatch)):

> If you return the same value from a Reducer Hook as the current state, React will bail out without rendering the children or firing effects.

Reply:
Nope, if we change the reference it doesnt trigger re-render.